workspace/notebook/py_sb_horizon_harmonised_appeal_has.json

```diff
@@ -0,0 +1,857 @@
+					"if \"IngestionDate\" not in hzn_src.columns:\r\n",
+					"    ing_col = (\r\n",
+					"        F.col(\"ingested_datetime\")\r\n",
+					"        if \"ingested_datetime\" in hzn_src.columns else\r\n",
+					"        F.col(\"expected_from\")\r\n",
+					"        if \"expected_from\" in hzn_src.columns else\r\n",
+					"        F.current_timestamp()\r\n",
+					"    )\r\n",
+					"    hzn_src = hzn_src.withColumn(\"IngestionDate\", ing_col)\r\n",
+					"\r\n",
+					"# Here we add the \"plumbing\" to look like SB rows (the final IsActive is recomputed by SCD-2 later)\r\n",
+					"hzn_aligned = (\r\n",
+					"    hzn_src\r\n",
+					"      .withColumn(\"AppealsHasID\", F.lit(None).cast(\"bigint\"))\r\n",
+					"      .withColumn(\"Migrated\", F.lit(\"0\"))\r\n",
+					"      .withColumn(\"ODTSourceSystem\", F.lit(\"HORIZON\"))\r\n",
+					"      .withColumn(\"ValidTo\", F.lit(None).cast(\"string\"))\r\n",
+					"      .withColumn(\"RowID\", F.lit(\"\"))\r\n",
+					"      .withColumn(\"IsActive\", F.lit(\"N\"))\r\n",
+					")\r\n",
+					"\r\n",
+					"# Here we align Horizon to SB schema (add missing columns and cast types)\r\n",
+					"sb_schema = service_bus_data.schema\r\n",
+					"for f in sb_schema:\r\n",
+					"    if f.name not in hzn_aligned.columns:\r\n",
+					"        hzn_aligned = hzn_aligned.withColumn(f.name, F.lit(None).cast(f.dataType))\r\n",
+					"    else:\r\n",
+					"        if str(hzn_aligned.schema[f.name].dataType) != str(f.dataType):\r\n",
+					"            hzn_aligned = hzn_aligned.withColumn(f.name, F.col(f.name).cast(f.dataType))\r\n",
+					"\r\n",
+					"# Here we drop the duplicate names case insensitively\r\n",
+					"seen = {}\r\n",
+					"for c in list(hzn_aligned.columns):\r\n",
+					"    lc = c.lower()\r\n",
+					"    if lc in seen:\r\n",
+					"        hzn_aligned = hzn_aligned.drop(c)\r\n",
+					"    else:\r\n",
+					"        seen[lc] = c\r\n",
+					"\r\n",
+					"# Here we project to the exact SB column order\r\n",
+					"proj = []\r\n",
+					"lower_map = {c.lower(): c for c in hzn_aligned.columns}\r\n",
+					"for f in sb_schema:\r\n",
+					"    if f.name in hzn_aligned.columns:\r\n",
+					"        proj.append(F.col(f.name).alias(f.name))\r\n",
+					"    elif f.name.lower() in lower_map:\r\n",
+					"        proj.append(F.col(lower_map[f.name.lower()]).cast(f.dataType).alias(f.name))\r\n",
+					"    else:\r\n",
+					"        proj.append(F.lit(None).cast(f.dataType).alias(f.name))\r\n",
+					"hzn_aligned = hzn_aligned.select(*proj)\r\n",
+					"\r\n",
+					"# Here we union Service bus to Stage\r\n",
+					"stg_df = service_bus_data.unionByName(hzn_aligned, allowMissingColumns=True).distinct()\r\n",
+					"\r\n",
+					"# Here we enforce non null key\r\n",
+					"null_key_cnt = stg_df.filter(F.col(\"caseReference\").isNull()).count()\r\n",
+					"print(f\"Stage rows with NULL caseReference: {null_key_cnt}\")\r\n",
+					"stg_df = stg_df.filter(F.col(\"caseReference\").isNotNull())"
```

Comment:
This bit seems redundant? Also the print statement could impact performance due to having to force pyspark evaluation